### DISCUSSION TITLE
Zod validations for undefined and skipped validations

### DISCUSSION THREAD

**Discussion by lennerd** [edmundhung/conform#488] [Ideas]
I like the way, conform utilises Zod error messages to skip validation in certain cases (should be validated on the server or should be skipped at this point). Not a big fan of the [code examples](https://conform.guide/validation#async-validation) though.

After I've experimented for a while I found a nice way of having custom Zod schemas to do the same thing: 

``` ts
import { conformZodMessage} from "@conform-to/zod";
import { z } from "zod";

export const validationUndefined = z.custom<never>(() => false, {
  message: conformZodMessage.VALIDATION_UNDEFINED,
  fatal: true,
});

export const validationSkipped = z.custom<never>(() => false, {
  message: conformZodMessage.VALIDATION_SKIPPED,
  fatal: false,
});
```

Than you can have a schema factory function like this:

``` ts
import { z, ZodType } from "zod";

export function createStaffSchema(options: { uniqueStaffName: ZodType<string> }) {
  return z.object({
    id: z.string().uuid(),
    name: z.string().trim().min(1).pipe(options.uniqueStaffName),
    role: z.enum(["STAFF", "ADMIN"]),
  });
}
```

``` ts
// On the client
createStaffSchema({ uniqueStaffName: validationUndefined });
```

``` ts
// On the server
const uniqueStaffName = z.string().refine(
  async (name) => {
    // Ask database if the name is unique
  },
  { message: "Staff name is not unique" }
);

createStaffSchema({ uniqueStaffName });
```

This moves the distinction for skipping some validation to the caller side while also a more composable way of combining validations.

Maybe this is something we can add to conform? Let me know what you think!

**Comment by edmundhung** [comment]
Thanks for sharing your idea! I agree that the current solution is far from ideal and what you suggested is interesting.

How would you handle `validationSkipped` based on the intent?

**Reply by lennerd** [reply]
Let's see:

``` ts
export async function action({ request }: ActionFunctionArgs) {
  const formData = await request.formData();
  const submission = await parseWithZod(formData, {
    schema: (intent) =>
      createStaffSchema({
        uniqueStaffName:
          intent === null || (intent.type === "validate" && intent.payload.name === "name")
            ? uniqueStaffName
            : validationSkipped,
      }),
  });

  // ...
}
```

We now indentify the intent outside the schema factory function which makes it more reusable and less coupled to conform.

What do you think?

**Reply by lennerd** [reply]
I'm on the server, so I use the `uniqueStaffName` schema to validate the name. Fixed the example.
